source/extensions/filters/udp/dns_filter/dns_parser.h

```diff
@@ -0,0 +1,192 @@
+                  const uint16_t rec_class, const uint32_t ttl,
+                  Network::Address::InstanceConstSharedPtr ipaddr)
+      : BaseDnsRecord(id, query_name, rec_type, rec_class), ttl_(ttl), ip_addr_(ipaddr) {}
+
+  ~DnsAnswerRecord() override = default;
+  void serialize(Buffer::OwnedImpl& output) override { UNREFERENCED_PARAMETER(output); }
+
+  const uint32_t ttl_;
+  Network::Address::InstanceConstSharedPtr ip_addr_;
+};
+
+using DnsAnswerRecordPtr = std::unique_ptr<DnsAnswerRecord>;
+using DnsAnswerMap = absl::flat_hash_map<std::string, std::list<DnsAnswerRecordPtr>>;
+
+enum class DnsQueryParseState {
+  Init = 0,
+  Flags,      // 2 bytes
+  Questions,  // 2 bytes
+  Answers,    // 2 bytes
+  Authority,  // 2 bytes
+  Authority2, // 2 bytes
+  Finish
+};
+
+/**
+ * This class orchestrates parsing a DNS query and building the response to be sent to a client.
+ */
+class DnsMessageParser : public Logger::Loggable<Logger::Id::filter> {
+
+public:
+  DnsMessageParser() = default;
+  ~DnsMessageParser() = default;
+
+  /**
+   * @param buffer a reference to the incoming request object received by the listener
+   * @return bool true if all DNS records and flags were successfully parsed from the buffer
+   */
+  bool parseDnsObject(const Buffer::InstancePtr& buffer);
+
+  /**
+   * @brief parse a single query record from a client request
+   *
+   * @param buffer a reference to the incoming request object received by the listener
+   * @param offset the buffer offset at which parsing is to begin. This parameter is updated when
+   * one record is parsed from the buffer and returned to the caller.
+   * @return DnsQueryRecordPtr a pointer to a DnsQueryRecord object containing all query data parsed
+   * from the buffer
+   */
+  DnsQueryRecordPtr parseDnsQueryRecord(const Buffer::InstancePtr& buffer, uint64_t* offset);
+
+  /**
+   * @return a reference to a list of queries parsed from a client request
+   */
+  const DnsQueryMap& getActiveQueryRecords() { return queries_; }
+
+  /**
+   * @return the current query transaction ID being handled
+   */
+  uint16_t getCurrentQueryId() const { return active_transactions_.front(); }
+
+  /**
+   * @return uint16_t the response code flag value from a parsed dns object
+   */
+  uint16_t getQueryResponseCode() { return static_cast<uint16_t>(incoming_.flags.rcode); }
+
+  /**
+   * Reset the internal state of the class.
+   */
+  void reset() {
+    active_transactions_.clear();
+    queries_.clear();
+  }
+
+private:
+  const std::string parseDnsNameRecord(const Buffer::InstancePtr& buffer, uint64_t* available_bytes,
+                                       uint64_t* name_offset);
+
+  /**
+   * @brief updates a map associating a query id with a list of DnsQueryRecord pointers
+   *
+   * @param rec the answer record that is to be added to the answer list
+   */
+  void storeQueryRecord(DnsQueryRecordPtr rec);
+
+  struct DnsHeader incoming_;
+  std::deque<uint16_t> active_transactions_;
```

Comment:
I didn't follow why this is required?

Reply:
The global active_transactions_  is not needed anymore. Correct?

Reply:
This was removed.